load topology for each device

bardliao opened the pull request:
Get device information from dai links. load topology for each device.

This should not impact the existing devices.

bardliao commented:
SOFCI TEST

bardliao committed:
ASoC: SOF: topology: allocate string for tuples

In preparation to handle tuples from multiple topologies, duplicate the
tuple string value by allocating memory for it.

Signed-off-by: Bard Liao <yung-chuan.liao@linux.intel.com>

ujfalusi commented on sound/soc/sof/topology.c:
```diff
@@ -571,7 +571,9 @@ static int sof_copy_tuples(struct snd_sof_dev *sdev, struct snd_soc_tplg_vendor_
-					tuples[*num_copied_tuples].value.s = elem->string;
+					tuples[*num_copied_tuples].value.s =
+						devm_kasprintf(sdev->dev, GFP_KERNEL,
+							       "%s", elem->string);
```
you need to check for NULL pointer.

ujfalusi commented on sound/soc/sof/topology.c:
```diff
@@ -2462,15 +2463,161 @@ static const struct snd_soc_tplg_ops sof_dspless_tplg_ops = {
+			return -EINVAL;
+		}
+
+		release_firmware(fw);
```
you can move the release_firmware() before checking for the ret

ujfalusi commented on sound/soc/sof/topology.c:
```diff
@@ -2462,15 +2463,161 @@ static const struct snd_soc_tplg_ops sof_dspless_tplg_ops = {
+		if (tplg_mask & BIT(tplg_dev))
+			continue;
+		tplg_mask |= BIT(tplg_dev);
+		tplg_files[tplg_num].be_id = dai_link->id;
```
you need to check for tplg_num < MAX_TPLG_NUM for safety!

ujfalusi commented on sound/soc/sof/topology.c:
```diff
@@ -2462,15 +2463,161 @@ static const struct snd_soc_tplg_ops sof_dspless_tplg_ops = {
+		ret = request_firmware(&fw, tplg_files[i].file, scomp->dev);
+		if (ret < 0) {
+			if (i == 0) {
+				dev_dbg(scomp->dev,"Fail back to %s\n", tplg_name);
```
If it was a topology w/ DMIC then you have dropped the -2ch in a previous loop, you are not going to load the topology you supposed to be loading.

bardliao replied:
Yes, you are right. I need to restore the original topology name.

ujfalusi commented on sound/soc/sof/topology.c:
```diff
@@ -2462,15 +2463,161 @@ static const struct snd_soc_tplg_ops sof_dspless_tplg_ops = {
+				tplg_name = strremove(tplg_name, "-2ch");
+			} else if (strstr(file, "-4ch")) {
+				tplg_device = "dmic-4ch";
+				tplg_name = strremove(tplg_name, "-4ch");
```
why are you dropping the 2ch/4ch from the original name?

bardliao replied:
Because I would like to load the topology without dmic. For example, sof-hda-generic-2ch.tplg -> sof-hda-generic.tplg or similar.

ujfalusi replied:
But the tplg_name is used only for loading the fallback, legacy, monolithic topology, in which case the fragments are not loaded. On a machine which have DMIC, you will load the topology w/o DMIC

bardliao replied:
Not exactly, tplg_name topology will be also loaded if not all dai links are handled. For example, if the monolithic topology is `sof-lnl-rt711-4ch`, then `sof-lnl-rt711` and `sof-lnl-dmic-4ch-id5` will be loaded

ujfalusi replied:
right, but if something fails in between and we need to fallback to the monolithic topology then at line 2641 you will have `sof-lnl-rt711.tplg` instead the correct monolithic `sof-lnl-rt711-4ch.tplg`, or am I missing something?

bardliao replied:
> right, but if something fails in between and we need to fallback to the monolithic topology then at line 2641 you will have `sof-lnl-rt711.tplg` instead the correct monolithic `sof-lnl-rt711-4ch.tplg`, or am I missing something?

Yeah, I will restore the tplg file name when it fails back to the monolithic topology, thanks.

ujfalusi commented on sound/soc/sof/topology.c:
```diff
@@ -2462,15 +2463,161 @@ static const struct snd_soc_tplg_ops sof_dspless_tplg_ops = {
+		} else {
+			/* The dai link is not supported by sperated tplg yet */
+			load_default_tplg = true;
+			continue;
```
why continue?

bardliao replied:
There is no supported topology for the dai link, but we want to load the topology for other dai links. Then the not handled dai links will be handled at line 2620 when the renamed topology is loaded.

ujfalusi commented on sound/soc/sof/topology.c:
```diff
@@ -2462,15 +2463,161 @@ static const struct snd_soc_tplg_ops sof_dspless_tplg_ops = {
+			if (strstr(file, "hda-generic")) {
+				tplg_device = "idisp";
+			} else {
+				tplg_device = "sdca-hdmi";
```
this is HDMI audio using SDW link or HDMI audio using HDA link, but used with SDW machine?

bardliao replied:
This is the HDMI audio used with SDW codecs. I use different name for different codec interfaces is because that the existing topologies use different PCM id for different codec interfaces. For example, the HDMI PCM ids are 5,6,7 for SDW machines and 3,4,5 for HDA machines.
```
00-00: Jack Out (*) :  : playback 1
00-01: Jack In (*) :  : capture 1
00-02: Speaker (*) :  : playback 1
00-03: Amp feedback (*) :  : capture 1
00-04: Microphone (*) :  : capture 1
00-05: HDMI1 (*) :  : playback 1
00-06: HDMI2 (*) :  : playback 1
00-07: HDMI3 (*) :  : playback 1
00-31: Deepbuffer Jack Out (*) :  : playback 1
```
```
00-00: HDA Analog (*) :  : playback 1 : capture 1
00-03: HDMI1 (*) :  : playback 1
00-04: HDMI2 (*) :  : playback 1
00-05: HDMI3 (*) :  : playback 1
00-06: DMIC Raw (*) :  : capture 1
00-31: Deepbuffer HDA Analog (*) :  : playback 1
```

ujfalusi commented on sound/soc/sof/topology.c:
```diff
@@ -2462,15 +2463,161 @@ static const struct snd_soc_tplg_ops sof_dspless_tplg_ops = {
+		} else if (strstr(dai_link->name, "iDisp")) {
+			tplg_dev = TPLG_DEVICE_HDMI;
+			if (strstr(file, "hda-generic")) {
+				tplg_device = "idisp";
```
If I read the function right, this is not going to be used as the hda-generic will fall under `load_default_tplg = true`, the HDA dai_link->name is not handled?

bardliao replied:
`load_default_tplg = true` means there are still some remaining pcms that are not loaded in the sparated topologies. And we need to load the renamed topology at line 2620.

ujfalusi replied:
but can you load the idisp.tplg and the sof-hda-generic.tplg which also contains idisp definitions?

bardliao replied:
> but can you load the idisp.tplg and the sof-hda-generic.tplg which also contains idisp definitions?

No, so we need to make sure there is no idisp in sof-hda-generic.tplg.

ujfalusi commented on sound/soc/sof/topology.c:
```diff
@@ -2462,15 +2463,161 @@ static const struct snd_soc_tplg_ops sof_dspless_tplg_ops = {
+						     "sdca-%damp", dai_link->num_cpus);
+		} else if (strstr(dai_link->name, "SmartMic")) {
+			tplg_dev = TPLG_DEVICE_SDW_MIC;
+			tplg_device = "sdca-mic";
```
Here and for the other checks: we need to update this match list every time a new component got introduced to products and we have new tplg fragment?
Is it going to scale?

bardliao replied:
We need to update this match list when a new dai links got introduced. Given that we didn't update dai links in our machine drivers. I think we will only update the match list occasionally.

ujfalusi commented on sound/soc/sof/topology.c:
```diff
@@ -2462,15 +2463,161 @@ static const struct snd_soc_tplg_ops sof_dspless_tplg_ops = {
+						    tplg_files[i].device,
+						    tplg_files[i].be_id);
+		dev_dbg(scomp->dev, "Requesting %d %s\n", i, tplg_files[i].file);
+		ret = request_firmware(&fw, tplg_files[i].file, scomp->dev);
```
the `sof_create_ipc_file_profile()` will check if the firmware and the topology file (the monolithic one!) is in place.
The means that you always need to have the monolithic tplg _and_ the split components installed, otherwise the profile is rejected and if not fallback IPC version is available then we will fail.

I have a feeling that this split tplg handling has to touch the fw-file-profile.c as well.

bardliao replied:
Yeah, I am thinking this open, too. My idea is that we don't touch the monolithic topology. I.e. the monolithic topology will coexist with the split topologies. So that sof_create_ipc_file_profile() will still valid.

ujfalusi commented on sound/soc/sof/topology.c:
```diff
@@ -2462,15 +2463,161 @@ static const struct snd_soc_tplg_ops sof_dspless_tplg_ops = {
+						    "%s/sof-%s-%s-id%d.tplg",
+						    sof_pdata->tplg_filename_prefix, platform,
+						    tplg_files[i].device,
+						    tplg_files[i].be_id);
```
a clean documentation as commit message and comments to explain clearly the expected file naming and also an update to sof-docs must be done.

bardliao replied:
Totally agree.

ujfalusi replied:
have you thought about the documentation?

bardliao replied:
I almost forget the document. I will do it next week.

bardliao replied:
Sorry for the delay. @ujfalusi The sof-docs PR is https://github.com/thesofproject/sof-docs/pull/507

bardliao committed:
ASoC: SOF: topology: don't convert error code

No need to convert the return value of snd_soc_tplg_component_load().

Signed-off-by: Bard Liao <yung-chuan.liao@linux.intel.com>

ujfalusi commented on sound/soc/sof/topology.c:
```diff
@@ -2489,7 +2489,6 @@ int snd_sof_load_topology(struct snd_soc_component *scomp, const char *file)
 		dev_err(scomp->dev, "error: tplg component load failed %d\n",
 			ret);
-		ret = -EINVAL;
 	}
```
you can also drop the curly brackets

ujfalusi commented on sound/soc/sof/topology.c:
```diff
@@ -2464,15 +2465,180 @@ static const struct snd_soc_tplg_ops sof_dspless_tplg_ops = {
 	.bytes_ext_ops_count = ARRAY_SIZE(sof_dspless_bytes_ext_ops),
 };
 
+#define MAX_TPLG_NUM 5
```
Add TPLG_DEVICE_MAX at the end of the enum?

bardliao replied:
Yes, it would be better.

ujfalusi commented on sound/soc/sof/topology.c:
```diff
@@ -2464,15 +2465,180 @@ static const struct snd_soc_tplg_ops sof_dspless_tplg_ops = {
+
+static bool is_platform_support_separated_tplg(const char *platform)
+{
+	if (!strcmp(platform, "mtl") || !strcmp(platform, "lnl") || !strcmp(platform, "ptl"))
```
this is not going to scale... You are already missing ARL...
I would allow this for SOF_IPC4, I guess you handle the case when the topology files are not split, so it should just fallback?

ujfalusi replied:
I guess you want to exclude the sof-hda on top of platform, right?
if !ipc4 || hda -> false ?
But this includes I2S machines also... Not sure what rule you want to make here.

bardliao replied:
My first thought is to only list the platforms that we already have the topology support. And "HDA" is one of the "platform" because we use sof-hda-generic.tplg for all platforms. Yeah, we can remove this check and fallback if the driver doesn't find the topology files.

ujfalusi commented on sound/soc/sof/topology.c:
```diff
@@ -2464,15 +2465,180 @@ static const struct snd_soc_tplg_ops sof_dspless_tplg_ops = {
+		} else if (strstr(dai_link->name, "SmartAmp")) {
+			tplg_dev = TPLG_DEVICE_SDW_AMP;
+			tplg_device = devm_kasprintf(sdev->dev, GFP_KERNEL,
+						     "sdca-%damp", dai_link->num_cpus);
```
can there be multiple SmartAmp link? You only going to record the first one. The naming of the device is also interesting, it is postfixed with a number of CPU dais?

bardliao replied:
We don't support multiple SmartAmp link in the machine driver. And, yes, postfixing with a number of CPU dais is intentional. We need to set NUM_SDW_AMP_LINKS in the topology.

ujfalusi commented on sound/soc/sof/topology.c:
```diff
@@ -2464,15 +2465,180 @@ static const struct snd_soc_tplg_ops sof_dspless_tplg_ops = {
+
+	for (i = 0; i < tplg_num; i++) {
+		tplg_files[i].file = devm_kasprintf(sdev->dev, GFP_KERNEL,
+						    "%s/sof-%s-%s-id%d.tplg",
```
How much permutations we are going to have for LNL dmic-2ch for the be_id fro example?

does sof-lnl-dmic-2ch-id2 compatible with non SDW cards (HDA, nocodec, I2S) or it is wired for SDW cards?

bardliao replied:
I think it is compatible with non SDW cards.

ujfalusi commented on sound/soc/sof/topology.c:
```diff
@@ -2464,15 +2465,180 @@ static const struct snd_soc_tplg_ops sof_dspless_tplg_ops = {
+		if (tplg_mask & BIT(tplg_dev))
+			continue;
+		tplg_mask |= BIT(tplg_dev);
+		tplg_files[tplg_num].be_id = dai_link->id;
```
Normally we have 3 display audio PCMs, you record the first one only, how does this be_id (used in generating the topology fragment name) is related? The HDMI fragment will have all PCMs, or?

bardliao replied:
Currently, the topology will always have 3 display audio PCMs. The be_id represents the first ID. For example, the IDs are 5, 6, and 7 for sof-arl-sdca-hdmi-id5.tplg

ujfalusi commented on sound/soc/sof/topology.c:
```diff
@@ -2464,15 +2465,180 @@ static const struct snd_soc_tplg_ops sof_dspless_tplg_ops = {
+	TPLG_DEVICE_SDW_JACK,
+	TPLG_DEVICE_SDW_AMP,
+	TPLG_DEVICE_SDW_MIC,
+	TPLG_DEVICE_HOST_DMIC,
```
This is awkwardly named, I think you meant INTEL_DMIC (conencted to PCH-DMIC)?

bardliao replied:
I will change it to TPLG_DEVICE_PCH_DMIC, thanks

ujfalusi commented on sound/soc/sof/topology.c:
```diff
@@ -2464,15 +2465,180 @@ static const struct snd_soc_tplg_ops sof_dspless_tplg_ops = {
+			tplg_dev = TPLG_DEVICE_HOST_DMIC;
+		} else if (strstr(dai_link->name, "iDisp")) {
+			tplg_dev = TPLG_DEVICE_HDMI;
+			tplg_device = "sdca-hdmi";
```
I know that you have excluded generic HDA, but things might fall through the cracks and picks this as HDMI where the PCM device id range is not 5,6,7. Which I suppose this fragment will provide.

How this naming going to scale with display audio serviced by SDW?

I'm not sure how this can be handled, but if we let HDA generic into \this logic, it is going to fail.

bardliao replied:
I added the be_id in the topology name, so that the PCM device id can be any number.

ujfalusi commented on sound/soc/sof/topology.c:
```diff
@@ -2464,15 +2465,180 @@ static const struct snd_soc_tplg_ops sof_dspless_tplg_ops = {
+				goto legacy_tplg;
+			}
+
+			dev_err(scomp->dev, "error: tplg request firmware %s failed err: %d\n",
```
Please don't use 'error:' in new prints.

ujfalusi commented on sound/soc/sof/topology.c:
```diff
@@ -2464,15 +2465,180 @@ static const struct snd_soc_tplg_ops sof_dspless_tplg_ops = {
+			return -ENOMEM;
+
+		dev_dbg(scomp->dev, "Requesting %d %s\n", i, tplg_files[i].file);
+		ret = request_firmware(&fw, tplg_files[i].file, scomp->dev);
```
firmware_request_nowarn() to handle the fallback in case the fragmented tplg is not present, but what will happen if you load two fragment thn the third is missing and try to load the monolithic over it?

I'm sure that this is going to happen.

ujfalusi replied:
and the original name is altered, the -2/4ch is dropped, so we are not going to load the right tplg in some cases.

bardliao replied:
> firmware_request_nowarn() to handle the fallback in case the fragmented tplg is not present, but what will happen if you load two fragment thn the third is missing and try to load the monolithic over it?
> 

It will return error if the second or later topology is not present. My idea is that if we can't find the first fragmented topology, it means we don't have the required version of topologies, then we need to fall back to use the original topology. However, if we find the first fragmented topology, it means we use the new version of topology, and there should be something wrong if we don't find any required fragmented topology.

> and the original name is altered, the -2/4ch is dropped, so we are not going to load the right tplg in some cases.

I will restore the file name in this case.

ujfalusi commented on sound/soc/sof/topology.c:
```diff
@@ -2464,15 +2465,180 @@ static const struct snd_soc_tplg_ops sof_dspless_tplg_ops = {
+
+	for (i = 0; i < tplg_num; i++) {
+		tplg_files[i].file = devm_kasprintf(sdev->dev, GFP_KERNEL,
+						    "%s/sof-%s-%s-id%d.tplg",
```
Another question: can a `sof-sdca-jack.tplg` work on all platforms, do we really need the mtl/lnl/ptl/arl/etc variants?

bardliao replied:
Good questions. We set 'PLATFORM' in the topology, but we don't set any variables for SDW. So, I think we can remove the platform name for the SDW topologies. It will be easier to implement the topologies.

bardliao commented:
The sof-docs PR is https://github.com/thesofproject/sof-docs/pull/507 and topology PR is https://github.com/thesofproject/sof/pull/9668. @ujfalusi @ranj063 @lgirdwood Could you review? Thanks.

ujfalusi commented on sound/soc/sof/topology.c:
```diff
@@ -2464,15 +2465,198 @@ static const struct snd_soc_tplg_ops sof_dspless_tplg_ops = {
+};
+
+/* The code is from https://stackoverflow.com/questions/47116974/remove-a-substring-from-a-string-in-c */
+static char *strremove(char *str, const char *sub)
```
I'm not sure if we want code from 'shady' places in kernel ;)

bardliao replied:
On the second thought, the current PR only supports split topologies when all devices are covered by the split topologies. There is no needed to change the original tplg file name. And we don't need *strremove() in the commit.

ujfalusi commented on sound/soc/sof/topology.c:
```diff
@@ -2464,15 +2465,198 @@ static const struct snd_soc_tplg_ops sof_dspless_tplg_ops = {
 	dev_dbg(scomp->dev, "loading topology:%s\n", file);
 
-	ret = request_firmware(&fw, file, scomp->dev);
+	tplg_name = devm_kasprintf(sdev->dev, GFP_KERNEL, "%s", file);
```
is this needs to be resource managed?

bardliao replied:
Fixed.

ujfalusi commented on sound/soc/sof/topology.c:
```diff
@@ -2464,15 +2465,198 @@ static const struct snd_soc_tplg_ops sof_dspless_tplg_ops = {
+	if (sdev->pdata->ipc_type == SOF_IPC_TYPE_3)
+		goto legacy_tplg;
+
+	ret = sscanf(sof_pdata->tplg_filename, "sof-%3s-*.tplg", platform);
```
So we _must_ always have a monolithic topology present in filesystem? During probe we check for existence of the topology file in sof_test_topology_file() and we will fail if it is missing.

bardliao replied:
Yes, we will create the monolithic topology. It is also for the old kernel that doesn't support split topologies.

ranj063 replied:
@bardliao I have a couple of fundamental problems with your solution. 
1. We start off as if there's only 1 topology file for each machine and then you come here in this function and decide thats not correct and then figure out what all sub-topologies to load. What does it mean "user will create topology for single device...". Its more like the user will create multiple generic mini topologies that are applicable across platforms isnt it?
2. Secondly, this is very Soundwire-specific isnt it? Can we make it a bit more generic so that it applies to all machines, SDW, I2S, HDA etc? We can have a coarse-grained split based on DAI types i.e SDW/I2S/HDA + HDMI + DMIC and a fine-grained split for the first part based onthe disco table for the SDW bits. This way we can do this change in 2 steps, first the split by DAI types and then isolate the SDW disco-specific split as a second step so it makes it more intuitive for everyone to understand.

ujfalusi commented on sound/soc/sof/topology.c:
```diff
@@ -2464,15 +2465,198 @@ static const struct snd_soc_tplg_ops sof_dspless_tplg_ops = {
 	const struct firmware *fw;
+	bool load_default_tplg = false;
+	unsigned long tplg_mask = 0;
+	char platform[4];
```
"tgl-h" needs 6 char space...

bardliao replied:
But the PLATFORM name is "tgl" in the topology.

ujfalusi replied:
right, but this is a generic code, you cannot assume that every vendor's every topology family will use three letters.

ranj063 replied:
no magic numbers like 4 please

ujfalusi commented on sound/soc/sof/topology.c:
```diff
@@ -2464,15 +2465,198 @@ static const struct snd_soc_tplg_ops sof_dspless_tplg_ops = {
 int snd_sof_load_topology(struct snd_soc_component *scomp, const char *file)
 {
 	struct snd_sof_dev *sdev = snd_soc_component_get_drvdata(scomp);
+	struct topology_file tplg_files[TPLG_DEVICE_MAX];
```
We can ever have 1 instance providing a function? no two jack/amp/mic/dmic/hdmi ? only ever one from each?

ujfalusi replied:
think about karaoke devices with multiple jacks/mics for example.

bardliao replied:
The idea is to have a topology for a function. Like we have 3 hdmi pcms, but we only have 1 hdmi topology.

ujfalusi reviewed (commented):
@bardliao, I will continue on Monday, sorry for the delay.

kv2019i commented on sound/soc/sof/topology.c:
```diff
@@ -2464,15 +2465,198 @@ static const struct snd_soc_tplg_ops sof_dspless_tplg_ops = {
+	if (ret != 1)
+		goto legacy_tplg;
+
+	for_each_card_prelinks(scomp->card, i, dai_link) {
```
This seems very cool and potential big save in tplg combination maintainance! 

One thing about this snd_sof_load_topology() logic is that how applicable this is to non-Intel SOF platforms? SDCA is in theory ok (same for all vendors), but concepts like "iDisp" and cards name lie "sof-hda-dsp" is already fairly vendor specific, while this function has been agnostic before. Should this be branched out via ops to allow for variation between vendors? I guess not all needs to be done immediately, but have you thought about this?

bardliao replied:
My first idea is that other vendors can add their link_name and card name to snd_sof_load_topology(). But yeah, using ops is more structured. Let me think more about it.

ranj063 commented on sound/soc/sof/topology.c:
```diff
@@ -571,7 +571,11 @@ static int sof_copy_tuples(struct snd_sof_dev *sdev, struct snd_soc_tplg_vendor_
 						continue;
 
 					tuples[*num_copied_tuples].token = tokens[j].token;
-					tuples[*num_copied_tuples].value.s = elem->string;
```
I'd change the commit message to say: "In preparation to handle tuples from multiple topologies, duplicate the tuple string value by allocating memory for it..."

bardliao commented:
@ranj063 
1. I didn't say that 1 topology file for each machine is not correct. But loading sub-topologies can significantly reduce our effort to enable a new codec configuration. We added more and more quirks to decide what endpoints are used in the machine. And create topologies for every single audio configuration. Those are not needed if we could detect what dai links should be created and what sub-topologies should be loaded. And, yes, it is more like that user can create a topology just for one device type. Like speakers. And the topology is applicable across platforms.
2. Yes, but I hope we can have this feature ASAP as there are more and more quirks/SDW topologies coming. So that I decide to focus on the SDW cards first.

bardliao commented:
@ranj063 @kv2019i I will implement a `card_tplg_ops`. So that we can focus on the 'sof-soundwire' card, and it is easily to extend to other cards includes the non-Intel cards.

ranj063 commented:
> @ranj063 @kv2019i I will implement a `card_tplg_ops`. So that we can focus on the 'sof-soundwire' card, and it is easily to extend to other cards includes the non-Intel cards.

@bardliao sounds good. Do you want to convert this to draft until you have the other solution ready?

bardliao committed:
ASoC: SOF: topology: load multiple topologies

Currently, we always use single topology file to describe the widgets.
However, with SDCA, we want to be able to load sub-topologies based on
the supported device functions.

Signed-off-by: Bard Liao <yung-chuan.liao@linux.intel.com>

lyakh commented on sound/soc/sof/topology.c:
```diff
@@ -2474,14 +2475,71 @@ static const struct snd_soc_tplg_ops sof_dspless_tplg_ops = {
+
+		if (ret < 0) {
+			dev_err(scomp->dev, "tplg component load failed %d\n",
+				ret);
```
would fit in 1 line

lyakh commented on sound/soc/sof/card-tplg-ops.c:
```diff
@@ -0,0 +1,114 @@
+		tplg_mask |= BIT(tplg_dev);
+		switch (tplg_dev) {
+		case TPLG_DEVICE_PCH_DMIC:
+			(*tplg_files)[tplg_num] = kasprintf(GFP_KERNEL, "%s/sof-%s-%s-id%d.tplg",
```
you've allocated `*tplg_files` as just a single pointer, and here you're using it as a virtually endless (depending on the number of the cards) array. This seems like a memory corruption to me.

bardliao replied:
That will not be endless. It is in the `for_each_card_prelinks()` loop which is `(i) < (card)->num_links)`. So, the max number of `tplg_num` is `num_links`. And I allocate `*tplg_files = kcalloc(scomp->card->num_links, sizeof(char *), GFP_KERNEL);` before entering the loop.

lyakh replied:
@bardliao oh, you're right, my oversight, sorry

lyakh commented on sound/soc/sof/topology.c:
```diff
@@ -2474,14 +2475,71 @@ static const struct snd_soc_tplg_ops sof_dspless_tplg_ops = {
 
 	dev_dbg(scomp->dev, "loading topology:%s\n", file);
 
+	for (i = 0; card_ops[i].card_name; i++) {
```
since the `card_ops` array is `static`, it means, that you're only going to use it in this file. Then you don't need to use an empty element in it to indicate termination, just use `ARRAY_SIZE()`

lyakh replied:
...I cannot dismiss my own "request-for-change" here... I'll approve it then, just to override it, the other two comments aren't critical, although the one for the array size would be good to address

bardliao replied:
Thanks @lyakh I addressed the other comments. :)

lgirdwood commented on sound/soc/sof/card-tplg-ops.c:
```diff
@@ -0,0 +1,114 @@
+#include "sof-priv.h"
+#include "topology.h"
+
+enum tplg_device_id {
```
I guess these are SDCA devices types ? I guess we can incrementally reuse this or comment as we align with SDCA

bardliao replied:
> I guess these are SDCA devices types ? I guess we can incrementally reuse this or comment as we align with SDCA

They are tplg devices not just SDCA devices types. Currently, we will create one mini-topology for one tplg device.

lgirdwood commented on sound/soc/sof/card-tplg-ops.c:
```diff
@@ -0,0 +1,114 @@
+#define SDCA_DEVICE_MASK (BIT(TPLG_DEVICE_SDW_JACK) | BIT(TPLG_DEVICE_SDW_AMP) | \
+			  BIT(TPLG_DEVICE_SDW_MIC))
+
+#define PLATFORM_NAME_MAX 4
```
I think 4 is enough for Intel, but maybe not long enough for other vendors ?

bardliao replied:
I will rename it to `SOF_INTEL_PLATFORM_NAME_MAX` as it will be used by Intel cards only.

lgirdwood commented on sound/soc/sof/card-tplg-ops.c:
```diff
@@ -0,0 +1,114 @@
+	if (!(*tplg_files))
+		return -ENOMEM;
+
+	for_each_card_prelinks(scomp->card, i, dai_link) {
```
Can we comment the naming rules here.

bardliao replied:
sure

ranj063 commented on sound/soc/sof/topology.c:
```diff
@@ -2321,7 +2322,7 @@ static const struct snd_soc_tplg_ops sof_tplg_ops = {
 
 	/* completion - called at completion of firmware loading */
-	.complete	= sof_complete,
+	/* complete will be added in the last tplg ops */
```
what do you mean by the last topology here?

bardliao replied:
The latest tplg file that we load.

ujfalusi replied:
Can you be a bit more outspoken with:
```c
/*
   completion - called at completion of firmware loading
   It is going to be set to sof_complete before the last topology fragment is loaded
   to finalize the topology loading.
*/
.complete = NULL,
```

ranj063 commented on sound/soc/sof/topology.c:
```diff
@@ -2474,14 +2475,69 @@ static const struct snd_soc_tplg_ops sof_dspless_tplg_ops = {
+		goto legacy_tplg;
+
+	for (i = 0; i < tplg_num; i++) {
+		dev_dbg(scomp->dev, "Requesting %d %s\n", i, tplg_files[i]);
```
I think it makes better sense to move this down below just before loading the topology

bardliao replied:
Hmm, but I need to know which file are needed. We will not know the file name if the first tplg file is not in the device.

ranj063 commented on sound/soc/sof/topology.c:
```diff
@@ -2474,14 +2475,69 @@ static const struct snd_soc_tplg_ops sof_dspless_tplg_ops = {
+		if (ret < 0) {
+			if (i == 0) {
+				dev_dbg(scomp->dev, "Fail back to %s\n", file);
+				kfree(tplg_files[i]);
```
why are we freeing this here, this function doesnt seem to allocate it?

bardliao replied:
tplg_files[] are allocated in card_ops[i].get_tplg_files(). Moved after the `out:` label and added a comment to describe where they are allocated.

ranj063 commented on sound/soc/sof/topology.c:
```diff
@@ -2474,14 +2475,69 @@ static const struct snd_soc_tplg_ops sof_dspless_tplg_ops = {
+
+			dev_err(scomp->dev, "tplg request firmware %s failed err: %d\n",
+				tplg_files[i], ret);
+			kfree(tplg_files[i]);
```
same here

ranj063 commented on sound/soc/sof/topology.c:
```diff
@@ -2474,14 +2475,69 @@ static const struct snd_soc_tplg_ops sof_dspless_tplg_ops = {
+			kfree(tplg_files[i]);
+			goto out;
+		}
+		kfree(tplg_files[i]);
```
and here

ranj063 commented on sound/soc/sof/topology.c:
```diff
@@ -2474,14 +2475,69 @@ static const struct snd_soc_tplg_ops sof_dspless_tplg_ops = {
+
+		/* set complete = sof_complete if it is the last topology */
+		if (i == tplg_num - 1)
+			sof_tplg_ops.complete = sof_complete;
```
why should we restrict complete to only the last sub topology? All we do there is update the IPC structures, set the pipeline widget etc. Can we make this independent for each sub-tooplogy?

bardliao replied:
All controls/widgets are added in the sdev lists and they will be handled. We need to split them if we want to handle them separately. Consider the same *scomp is passed to the .complete ops, I don't know how to make it independent for each sub-tooplogy. And actually, I don't get the benefit of make it independently.

ranj063 replied:
instead of setting the complete op why not just simply call sof_complete() here after successfully loading the last topology?

ranj063 commented on sound/soc/sof/topology.c:
```diff
@@ -2474,14 +2475,69 @@ static const struct snd_soc_tplg_ops sof_dspless_tplg_ops = {
+		release_firmware(fw);
+
+		if (ret < 0) {
+			dev_err(scomp->dev, "tplg component load failed %d\n", ret);
```
the return value will be printed anyway, can we print the name of the topology that failed here?

bardliao replied:
Yes, done

ranj063 commented on sound/soc/sof/topology.c:
```diff
@@ -2474,14 +2475,69 @@ static const struct snd_soc_tplg_ops sof_dspless_tplg_ops = {
+		}
+	}
+	/* Load topology successfully, goto out */
+	goto out;
```
I think this part below should be condensed into the loop. Start with tplg_num = 1. Set the tplg_files[0] = file to begin with. Then you can remove the whole block below.

bardliao replied:
Good suggestion, done

ranj063 commented on sound/soc/sof/card-tplg-ops.c:
```diff
@@ -0,0 +1,120 @@
+// SPDX-License-Identifier: (GPL-2.0-only OR BSD-3-Clause)
```
@bardliao an alternative thought. Can we move this to the machine driver? I think it makes sense for this code to coexist along with the machine info and dailinks there.

bardliao replied:
I thought about it. However, the topology stuff is more like belong to component level instead of card level. Besides, we need the `*sof_pdata` data to construct the tplg file name.

ranj063 replied:
you've got tplg file name is available also in struct snd_soc_acpi_mach. I dont think you really need sof_pdata for this

ranj063 replied:
> I thought about it. However, the topology stuff is more like belong to component level instead of card level. Besides, we need the `*sof_pdata` data to construct the tplg file name.

I agree with topology stuff belonging to component but this has nothing to do with the topology stuff but rather the machine level stuff ie whether there's a mic, a speaker a dmic etc no?

bardliao replied:
@ranj063 I moved the ops to ` sound/soc/intel/common/`, and set the ops in each acpi_mach that needs it. I don't set the ops in sof_sdw.c because I can't access the same acpi_mach pointer in sof_sdw.c and SOF driver. The acpi_mach data is copied to the platform device not the pointer.

ranj063 commented:
@bardliao I still have some fundamental issues with this approach:
1. The card_tplg_ops seems like it should belong in the machine driver code and not the SOF core.
2. We end up parsing all the sub-topologies and storing all the info in sdev with no distinction that they come from different topologies.
3. Error handling: If a sub-topology load fails, you dont seem to unload the first few successfully loaded topologies

ranj063 commented on sound/soc/sof/topology.c:
```diff
@@ -2474,14 +2475,69 @@ static const struct snd_soc_tplg_ops sof_dspless_tplg_ops = {
+
+		if (ret < 0) {
+			dev_err(scomp->dev, "tplg component load failed %d\n", ret);
+			return ret;
```
this is incomplete. With a monolithic topology, an error would make sure to remove any successfully objects during topology loading. But with your approach, you leave the sub-topologies loaded

bardliao replied:
Added snd_soc_tplg_component_remove()

bardliao commented:
> @bardliao I still have some fundamental issues with this approach:
> 
> 1. The card_tplg_ops seems like it should belong in the machine driver code and not the SOF core.

IMHO, the whole topology stuff belongs to the component level instead of the card level. That's why it is in the SOF core.

> 2. We end up parsing all the sub-topologies and storing all the info in sdev with no distinction that they come from different topologies.

Yes, my point is that we just put the widgets in different topology files, but they all belong to the same audio system. IOW, they all work together no matter where they are from. I.e. The outcome should be identical no matter we load a monolithic topology or some sub-topologies.

> 3. Error handling: If a sub-topology load fails, you dont seem to unload the first few successfully loaded topologies

Yes, handled in the updated code.

bardliao committed:
ASoC: Intel: add sof_sdw_get_tplg_files ops

Add sof_sdw_get_tplg_files ops to get sub-topology file names for the
sof_sdw card.

Signed-off-by: Bard Liao <yung-chuan.liao@linux.intel.com>

bardliao committed:
ASoC: Intel: soc-acpi-intel-lnl-match: set get_function_tplg_files ops

The audio configs with multi-function SDCA codecs can use the
sof_sdw_get_tplg_files ops to get function topologies dynamically.

Signed-off-by: Bard Liao <yung-chuan.liao@linux.intel.com>

bardliao committed:
ASoC: Intel: soc-acpi-intel-arl-match: set get_function_tplg_files ops

The audio configs with multi-function SDCA codecs can use the
sof_sdw_get_tplg_files ops to get function topologies dynamically.

Signed-off-by: Bard Liao <yung-chuan.liao@linux.intel.com>

bardliao committed:
ASoC: Intel: soc-acpi-intel-ptl-match: add get_function_tplg_files ops

The audio configs with multi-function SDCA codecs can use the
sof_sdw_get_tplg_files ops to get function topologies dynamically.

Signed-off-by: Bard Liao <yung-chuan.liao@linux.intel.com>

bardliao committed:
ASoC: Intel: soc-acpi-intel-mtl-match: add get_function_tplg_files ops

The audio configs with multi-function SDCA codecs can use the
sof_sdw_get_tplg_files ops to get function topologies dynamically.

Signed-off-by: Bard Liao <yung-chuan.liao@linux.intel.com>

bardliao committed:
ASoC: soc-acpi: add get_function_tplg_files ops

We always use a single topology that contains all PCM devices belonging
to a machine configuration.
However, with SDCA, we want to be able to load function topologies based
on the supported device functions. This change is in preparation for
loading those function topologies.

Signed-off-by: Bard Liao <yung-chuan.liao@linux.intel.com>

ujfalusi commented on sound/soc/sof/topology.h:
```diff
@@ -0,0 +1,18 @@
+struct card_tplg_ops {
+	const char *card_name;
+	int (*get_tplg_files)(struct snd_soc_component *scomp, const char *file,
+			      const char ***tplg_files);
```
The return value must be documented, it is far from obvious from this that it returns -ERRNO or the number of files placed in tplg_files array.
Is 0 valid return value?

ujfalusi commented on sound/soc/sof/topology.c:
```diff
@@ -2474,37 +2475,93 @@ static const struct snd_soc_tplg_ops sof_dspless_tplg_ops = {
+	for (i = 0; i < ARRAY_SIZE(card_ops); i++) {
+		if (!strcmp(scomp->card->name, card_ops[i].card_name)) {
+			ret = card_ops[i].get_tplg_files(scomp, file, &tplg_files);
+			if (ret < 0)
```
you can use the `file_num` directly, but I would rename the *_num to *_cnt, for me the _num is more like pointing to a specific index and here we have count.
or num_files.

you can as well just get away with using the single `num_files` or `tplg_cnt`

bardliao replied:
I use` tplg_cnt` and `file_cnt` separately because the `tplg_cnt` can be changed when we fail back to the original topology. And we need to free the `file_cnt` number of `tplg_files[]`

ujfalusi commented on sound/soc/sof/topology.c:
```diff
@@ -2476,6 +2476,10 @@ static const struct snd_soc_tplg_ops sof_dspless_tplg_ops = {
 
 static const struct card_tplg_ops card_ops[] = {
+	{
+		.card_name = "sof-soundwire",
```
`sof-soundwire` is kind of reserved as "INTEL"-sof-soundwire, right? Can we add comment to spell this out?

ujfalusi commented on sound/soc/sof/card-tplg-ops.c:
```diff
@@ -0,0 +1,116 @@
+
+#define SOF_INTEL_PLATFORM_NAME_MAX 4
+
+int get_sof_sdw_tplg_files(struct snd_soc_component *scomp, const char *file,
```
add _intel_ in the function name, we only handle SOF_INTEL_PLATFORM_NAME_MAX long platform names.

ujfalusi commented on sound/soc/sof/card-tplg-ops.c:
```diff
@@ -0,0 +1,116 @@
+	TPLG_DEVICE_SDCA_JACK,
+	TPLG_DEVICE_SDCA_AMP,
+	TPLG_DEVICE_SDCA_MIC,
+	TPLG_DEVICE_PCH_DMIC,
```
TPLG_DEVICE_INTEL_PCH_DMIC
or
TPLG_DEVICE_SOC_DMIC

ujfalusi commented on sound/soc/sof/topology.h:
```diff
@@ -0,0 +1,18 @@
+ */
+
+#ifndef __SOUND_SOC_SOF_TOPOLOGY_H
+#define __SOUND_SOC_SOF_TOPOLOGY_H
```
I would rather create `card-tplg-ops.h` instead.

We might (_might_) be able to extend this to be used by the fw-file-profile.c and cut the reliance on the monolithic topology file even if we only expect to use a generic one.

But, but, I guess we can create a generic sof-<plat>-sdca-full-card.tplg and use this single one for all sdca setups...?

ujfalusi reviewed (commented):
@bardliao, I really like this version much better!
Few comments inside.

ranj063 commented on sound/soc/intel/common/soc-acpi-intel-get-tplg.c:
```diff
@@ -0,0 +1,118 @@
+	if (ret != 1)
+		return 0;
+
+	pr_err("bard: %s plarform: %s\n", card->name, platform);
```
remove?

ranj063 commented on sound/soc/intel/common/soc-acpi-intel-get-tplg.c:
```diff
@@ -0,0 +1,118 @@
+		} else {
+			/* The dai link is not supported by sperated tplg yet */
+			dev_dbg(card->dev,
+				"dai_link %s is not supported by sperated tplg yet\n",
```
typo seperated both lines but do we need the dbg statement at all?

bardliao replied:
Yes, it can remind us what dai links are still not supported by separated tplg.

ranj063 commented on sound/soc/sof/topology.c:
```diff
@@ -2477,34 +2482,87 @@ static const struct snd_soc_tplg_ops sof_dspless_tplg_ops = {
+out:
+	/* Free tplg_files[] which are allocated by card_ops[i].get_tplg_files() */
+	for (i = 0; i < file_cnt; i++)
+		kfree(tplg_files[i]);
```
This still looks awkward to free here. Why not use devm to allocate memory in the get_tplg_files op so that when the machine driver is removed due to this error, it will get freed automatically?

ranj063 commented on sound/soc/intel/common/soc-acpi-intel-get-tplg.c:
```diff
@@ -0,0 +1,118 @@
+	pr_err("bard: %s plarform: %s\n", card->name, platform);
+
+	for_each_card_prelinks(card, i, dai_link) {
+		char *tplg_device;
```
maybe change this to tplg_device_name or tplg_dev_name so that people dont confused why there's a tplg_dev and tplg_device

ranj063 commented on sound/soc/intel/common/soc-acpi-intel-get-tplg.c:
```diff
@@ -0,0 +1,118 @@
+
+	ret = sscanf(mach->sof_tplg_filename, "sof-%3s-*.tplg", platform);
+	if (ret != 1)
+		return 0;
```
isnt this as error? If not, maybe you can change the return type of this function to uint so that it is explicit

bardliao replied:
Yes, it is an error. Changed the return value and added a trace.

ranj063 commented on sound/soc/intel/common/soc-acpi-intel-get-tplg.c:
```diff
@@ -0,0 +1,118 @@
+			dev_dbg(card->dev,
+				"dai_link %s is not supported by sperated tplg yet\n",
+				dai_link->name);
+			return 0;
```
should we return here or continue?

bardliao replied:
We should return 0 so that snd_sof_load_topology() will use the monolithic topology.

ranj063 reviewed (commented):
@bardliao this looks a lot more intuitive. Thanks! Just a few more comments

ranj063 commented on sound/soc/sof/topology.c:
```diff
@@ -2477,34 +2482,87 @@ static const struct snd_soc_tplg_ops sof_dspless_tplg_ops = {
+		ret = firmware_request_nowarn(&fw, tplg_files[i], scomp->dev);
+		if (ret < 0) {
+			if (i == 0 && file_cnt) {
+				dev_dbg(scomp->dev, "Fail back to %s\n", file);
```
@bardliao you're assuming here that the monolithic topology file always exists which is true for the current platforms. But once we move to the split sub-topologies, we may not even have the monolithic topologies for future platforms right?

bardliao replied:
No, I didn't assume that the monolithic topology file always exists. I tested the return value of `request_firmware(&fw, file, scomp->dev);` and return error if request_firmware() failed. That is the same as current code.

ranj063 replied:
if you didnt assume there's a monolithic topology, what did you mean by fall back? and there's a typo in fallback

bardliao replied:
It means we tried the sub-topologies but failed. Now we are trying the monolithic topology. But we will know if the monolithic topology exists after calling request_firmware()

ranj063 commented on sound/soc/intel/common/soc-acpi-intel-get-tplg.c:
```diff
@@ -0,0 +1,118 @@
+		/*
+		 * The tplg file naming rule is sof-<platform>-<function>-id<BE id number>.tplg
+		 * where <platform> is only required for the DMIC function as the nhlt blob
+		 * is platform depentent.
```
typo dependent

ranj063 commented on sound/soc/sof/topology.c:
```diff
@@ -2477,34 +2482,80 @@ static const struct snd_soc_tplg_ops sof_dspless_tplg_ops = {
+		tplg_cnt = sof_pdata->machine->get_tplg_files(scomp->card, sof_pdata->machine,
+							      sof_pdata->tplg_filename_prefix,
+							      &tplg_files);
+		if (tplg_cnt < 0)
```
you need to free tplg_files in the error path

ranj063 commented on sound/soc/sof/topology.c:
```diff
@@ -2477,34 +2482,80 @@ static const struct snd_soc_tplg_ops sof_dspless_tplg_ops = {
+				}
+			} else {
+				if (i > 0)
+					snd_soc_tplg_component_remove(scomp);
```
actually after thinking about it, I think you dont need to do this at all. All of the aub-topologies are associated with the same component, so if a failure happens during loading for any of the sub-topologies, the component remove will handle all of the previously successfully loaded objects automatically. Maybe you can just add that comment here for clarity

ujfalusi commented on include/sound/soc-acpi.h:
```diff
@@ -193,6 +194,13 @@ struct snd_soc_acpi_link_adr {
  * @sof_tplg_filename: Sound Open Firmware topology file name, if enabled
  * @tplg_quirk_mask: quirks to select different topology files dynamically
+ * @get_tplg_files: to get sub-topology file names. The return value is the number
+ * 		    of the files.
```
or negative ERRNO, 0 is invalid/valid or has special meaning?

bardliao replied:
0 is valid and means the sub-topology doesn't exist. It is likely that the old version of sof-bin is used.

ujfalusi commented on sound/soc/sof/topology.c:
```diff
@@ -2477,34 +2479,82 @@ static const struct snd_soc_tplg_ops sof_dspless_tplg_ops = {
+						file, ret);
+					dev_err(scomp->dev,
+						"you may need to download the firmware from https://github.com/thesofproject/sof-bin/\n");
+					goto out;
```
this should not happen as the combined file has been checked for existence.

bardliao replied:
Hmm this is copied from the existing code
https://github.com/thesofproject/linux/blob/topic/sof-dev/sound/soc/sof/topology.c#L2485

ujfalusi commented on sound/soc/sof/topology.c:
```diff
@@ -2477,34 +2479,82 @@ static const struct snd_soc_tplg_ops sof_dspless_tplg_ops = {
-			ret);
+	for (i = 0; i < tplg_cnt; i++) {
+		dev_dbg(scomp->dev, "loading topology %d: %s\n", i, tplg_files[i]);
+		ret = firmware_request_nowarn(&fw, tplg_files[i], scomp->dev);
```
I think the get_tplg_files() should be checking the presence of the files or have a separate loop which only does that.
If any of the files are missing then fall-back to the combined file, which is known to exist as it was checked earlier during probe.

bardliao replied:
Ok, I will do the change.

ujfalusi commented on sound/soc/sof/fw-file-profile.c:
```diff
@@ -192,7 +192,8 @@ sof_file_profile_for_ipc_type(struct snd_sof_dev *sdev,
 
-	if (!ret)
+	/* No need to test topology file if get_tplg_files is set */
+	if (!ret && !plat_data->machine->get_tplg_files)
```
I'm not entirely sure about this, we must have standalone topology file for backwards compatibility and signaling the number of DMICs for example.
The code in topology.c can rely on the fact that this is in place for example.

bardliao replied:
My idea is that we may not release the standalone topology file in new sof-bin. And if the old sof-bin is used, we will still show the error when fail to request_firmware().

ujfalusi replied:
You need to have a match defined in sound/soc/intel/common/ and that includes a topology file field.
Are you planning to leave that to NULL or set it to 'don_t_care'?

and if you have the old sof-bin release which only have the combined file then you don't test it early and the fallback code can only fallback if the _first_ fragment is missing, otherwise it will hard fail.

bardliao replied:
> You need to have a match defined in sound/soc/intel/common/ and that includes a topology file field. Are you planning to leave that to NULL or set it to 'don_t_care'?

My plan is leave it to NULL

> 
> and if you have the old sof-bin release which only have the combined file then you don't test it early and the fallback code can only fallback if the _first_ fragment is missing, otherwise it will hard fail.

In this case, get_tplg_files() will return 0, and we will load the combined file directly. And it won't fallback as tplg_cnt == 1 in this case.

bardliao replied:
On the second thought, set a generic tplg to sof_tplg_filename in the match table will be much easier and seems make more sense.

ujfalusi reviewed (commented):
@bardliao, somehow this is now again a bit confusing.

I think we should keep the check in fw-profile for the combine topology file.
In the get_tplg_files callback (or in topology.c) we should check the existence of the per function fragments and if any of them is missing fall back to the combined one.

But in the fw-profile change the print for tplg name, as we might not load this at all. Hmm, so we should also print the topology files that we are intend to load if it is not the combined one to be able to see and troubleshoot if needed.

ranj063 commented:
> @bardliao, somehow this is now again a bit confusing.
> 
> I think we should keep the check in fw-profile for the combine topology file. In the get_tplg_files callback (or in topology.c) we should check the existence of the per function fragments and if any of them is missing fall back to the combined one.

@ujfalusi I think the point that is difficult to get from this PR is that once we have the sub-topologies shipped with sof-bin, we're going to have all fragments available. So checking if the first is available is enough. 
> 
> But in the fw-profile change the print for tplg name, as we might not load this at all. Hmm, so we should also print the topology files that we are intend to load if it is not the combined one to be able to see and troubleshoot if needed.
I think he does that in the PR as and when each of the topologies is loaded and it will complain when one of them fails to load.

ujfalusi commented on include/sound/soc-acpi.h:
```diff
@@ -193,6 +194,14 @@ struct snd_soc_acpi_link_adr {
  * @tplg_quirk_mask: quirks to select different topology files dynamically
+ * @get_tplg_files: to get sub-topology file names. The return value is the number
+ * 		    of the files or negative ERRNO. 0 is a valid value which means the
+ * 		    sub-topology doesn't exist.
```
nitpick: We could also just return the `mach->sof_tplg_filename` in tplg_files[0] if the split is not supported

bardliao replied:
I will leave it as it is as we don't pass the original file name to the callback function. Also, 
```
        if (!tplg_cnt) {
                tplg_files[0] = file;
                tplg_cnt = 1;
        }
```
looks intuitive to me.

ujfalusi commented on include/sound/soc-acpi.h:
```diff
@@ -193,6 +194,14 @@ struct snd_soc_acpi_link_adr {
  *  is not constant since this field may be updated at run-time
  * @sof_tplg_filename: Sound Open Firmware topology file name, if enabled
  * @tplg_quirk_mask: quirks to select different topology files dynamically
+ * @get_tplg_files: to get sub-topology file names. The return value is the number
```
this is optional callback, if specified then instead of the single `sof_tplg_filename` the callback will return the list of function topology files to be loaded.

I would also rename the `get_tplg_files` to `get_function_topology_files` to make things a bit cleaner ( I could not think a better name for soc-acpi-intel-get-tplg.c , perhaps sof-function-topology[-lib].c ?

bardliao replied:
Consider to the callback name length, I will rename `get_tplg_files` to `get_sub_tplg_files`.

ujfalusi commented on sound/soc/sof/topology.c:
```diff
@@ -2477,34 +2479,69 @@ static const struct snd_soc_tplg_ops sof_dspless_tplg_ops = {
-		ret = snd_soc_tplg_component_load(scomp, &sof_dspless_tplg_ops, fw);
-	else
-		ret = snd_soc_tplg_component_load(scomp, &sof_tplg_ops, fw);
+	if (!tplg_cnt) {
```
I would add a comment for the reader that we use the single topology file in case there is not get_function_topology_files callback or if that returns with 0 function topology.

ujfalusi commented on sound/soc/sof/topology.c:
```diff
@@ -2477,34 +2479,69 @@ static const struct snd_soc_tplg_ops sof_dspless_tplg_ops = {
-			ret);
+	for (i = 0; i < tplg_cnt; i++) {
+		dev_dbg(scomp->dev, "loading topology %d: %s\n", i, tplg_files[i]);
+		ret = firmware_request_nowarn(&fw, tplg_files[i], scomp->dev);
```
I think you can keep the `request_firmware()` use here, this should not fail anymore.

ujfalusi commented on sound/soc/sof/topology.c:
```diff
@@ -2477,34 +2479,69 @@ static const struct snd_soc_tplg_ops sof_dspless_tplg_ops = {
+	}
+
+	/* call sof_complete when topologies are loaded successfully */
+	ret = sof_complete(scomp);
```
oh, and this works? Cool!

ujfalusi commented on sound/soc/intel/common/soc-acpi-intel-get-tplg.c:
```diff
@@ -0,0 +1,135 @@
+	dev_dbg(card->dev, "tplg_mask %#lx tplg_num %d\n", tplg_mask, tplg_num);
+
+	/* Check presence of sub-topologies */
+	for (i = 0; i < tplg_num; i++) {
```
:+1:

ujfalusi commented on sound/soc/intel/common/soc-acpi-intel-get-tplg.c:
```diff
@@ -0,0 +1,135 @@
+#include <linux/firmware.h>
+#include <sound/soc.h>
+#include <sound/soc-acpi.h>
+#include "soc-acpi-intel-get-tplg.h"
```
probably a file rename if you agree with my previous comment.

ujfalusi reviewed (commented):
@bardliao, this is now really nice, I have few naming suggestions to make it more defined.

ujfalusi commented on include/sound/soc-acpi.h:
```diff
@@ -193,6 +194,15 @@ struct snd_soc_acpi_link_adr {
  *  is not constant since this field may be updated at run-time
  * @sof_tplg_filename: Sound Open Firmware topology file name, if enabled
  * @tplg_quirk_mask: quirks to select different topology files dynamically
+ * @get_sub_tplg_files: This is an optional callback, if specified then instead of
```
Yeah, but `sub` to what? There is a main tplg file which these are subs to?
`get_function_tplg_files` would describe what this does better.
For me the tplg sounds sloppy compared to topology, but if the concern is length, then it is fine.

bardliao replied:
Renamed to `get_function_tplg_files` now

ujfalusi commented on include/sound/soc-acpi.h:
```diff
@@ -193,6 +194,15 @@ struct snd_soc_acpi_link_adr {
+ * 	the single sof_tplg_filename the callback will return the list of function topology
+ * 	files to be loaded.
+ * 	Return value: The number of the files or negative ERRNO. 0 is a valid value which means
+ * 		      the sub-topology doesn't exist.
```
0 means that the single topology file should be used, no function topology split can be used on the machine.

ujfalusi commented on include/sound/soc-acpi.h:
```diff
@@ -193,6 +194,15 @@ struct snd_soc_acpi_link_adr {
  *  is not constant since this field may be updated at run-time
  * @sof_tplg_filename: Sound Open Firmware topology file name, if enabled
  * @tplg_quirk_mask: quirks to select different topology files dynamically
+ * @get_function_tplg_files: This is an optional callback, if specified then instead of
```
the commit message also needs to be changed.

ujfalusi commented on sound/soc/intel/common/soc-acpi-intel-lnl-match.c:
```diff
@@ -575,6 +576,7 @@ struct snd_soc_acpi_mach snd_soc_acpi_intel_lnl_sdw_machines[] = {
 		.links = lnl_cs42l43_l0,
 		.drv_name = "sof_sdw",
 		.sof_tplg_filename = "sof-lnl-cs42l43-l0.tplg",
+		.get_function_tplg_files = sof_sdw_get_tplg_files,
```
the commit messages for the match files also needs update s/get_tplg_files/get_function_tplg_files

ujfalusi replied:
commit message: s/get_tplg_function_files/get_function_tplg_files

ranj063 commented on sound/soc/sof/topology.c:
```diff
@@ -2477,34 +2479,75 @@ static const struct snd_soc_tplg_ops sof_dspless_tplg_ops = {
-	else
-		ret = snd_soc_tplg_component_load(scomp, &sof_tplg_ops, fw);
+	/*
+	 * The monolithic topology will be used if there is no get_sub_tplg_files
```
typo get_function_tplg_files

bardliao replied:
fixed

ranj063 reviewed (dismissed):
just one more typo @bardliao

ujfalusi commented on sound/soc/sof/topology.c:
```diff
@@ -2473,36 +2479,75 @@ static const struct snd_soc_tplg_ops sof_dspless_tplg_ops = {
-			ret);
-		ret = -EINVAL;
+	for (i = 0; i < tplg_cnt; i++) {
+		dev_info(scomp->dev, "loading topology %d: %s\n", i, tplg_files[i]);
```
One nitpick:
```c
if (tplg_cnt != 1)
	dev_info(scomp->dev, "loading function topology %d: %s\n", i, tplg_files[i]);
else
	dev_dbg(scomp->dev, "loading topology:%s\n", file);
```

In case of a single topology file we would print in debug level as we did before, but in case of using function topology fragment we print with info

ujfalusi replied:
Or even better:
```c
	if (tplg_cnt == 1)
		dev_dbg(scomp->dev, "loading topology:%s\n", file);
	else
		dev_info(scomp->dev, "Using function topologies instead %s\n", file);

	for (i = 0; i < tplg_cnt; i++) {
		if (tplg_cnt != 1)
			dev_info(scomp->dev, " loading funtion %d: %s\n", i, tplg_files[i]);
```

ujfalusi reviewed (commented):
@bardliao, thank you, this is pretty cool, could not wait to see it in

ujfalusi commented on sound/soc/sof/topology.c:
```diff
@@ -2473,36 +2479,80 @@ static const struct snd_soc_tplg_ops sof_dspless_tplg_ops = {
+	if (!tplg_cnt) {
+		tplg_files[0] = file;
+		tplg_cnt = 1;
+		dev_dbg(scomp->dev, "loading topology %d: %s\n", i, tplg_files[i]);
```
i is not initialized at this point yet.

ujfalusi replied:
and I would keep the original print of
```
dev_dbg(scomp->dev, "loading topology:%s\n", file);
```

ujfalusi reviewed (dismissed):
@bardliao, thank you, this looks really nice now, could not wait to see this deployed and in use!!

ranj063 commented:
@bardliao looks like there are some legit checkpatch errors. Can you please fix them?

ranj063 commented:
@ujfalusi can u please merge if you're good
